### ISSUE TITLE
Ordering of classes depends on order of function statement execution

### ISSUE THREAD

**Issue by oscarbenjamin** [mrocklin/multipledispatch#118]
I discovered this while looking into https://github.com/sympy/sympy/issues/21968

```python
from multipledispatch import dispatch

class A: pass

class B: pass

class C(A, B): pass

CALL_A = True

if CALL_A:
    # The A method will be called
    @dispatch(B)
    def f(b):
        print('B method')

    @dispatch(A)
    def f(a):
        print('A method')
else:
    # The B method will be called
    @dispatch(A)
    def f(a):
        print('A method')

    @dispatch(B)
    def f(b):
        print('B method')

f(C())
```
Depending on the order of the dispatch functions for A and B (as executed by the interpreter) different dispatch functions are called. This is problematic especially if those different functions are in different modules that could be imported in different orders.

Actually if it wasn't for the use of OrderedDict here
https://github.com/mrocklin/multipledispatch/blob/880e700c6fd5c17a4cd0fc0f14ec3c1fbf4255b7/multipledispatch/utils.py#L57
then the behaviour would be nondeterministic regardless of the order of definitions.

It seems unclear to me whether the behaviour of multiple dispatch as implemented is well specified. I had thought that the topological sort was enough to specify which dispatch should be called but I can see clearly now that it is not (even in the case of single dispatch).

Maybe a case like above should trigger an ambiguity warning/error.

**Comment by gwangjinkim** [comment]
I stumbled over the same problem.

It is possible, to let the dispatch work gradually from most specific to the inherited less specific cases, in my view.

**Comment by mrocklin** [comment]
> It is possible, to let the dispatch work gradually from most specific to the inherited less specific cases, in my view

Yes.  It does this.  The challenge posed above is that neither A nor B is more specific than the other, so it is unclear which should take precedence.  Today that precedence is determined by order of import.  If someone wants to propose a different tiebreaker (like the lexicographical ordering of strings) I'm open to it.  I think that ideally people would do a bit of research into what other systems (perhaps Julia) does and bring that information back here.

**Comment by oscarbenjamin** [comment]
Could this not be the tiebreaker:
```python
>>> C.mro()
[__main__.C, __main__.A, __main__.B, object]
```

**Comment by mrocklin** [comment]
It could be.  I'm curious what CPython is choosing as tiebreaker.  String
ordering?  Import ordering?

**Comment by oscarbenjamin** [comment]
http://python-history.blogspot.com/2010/06/method-resolution-order.html

It is the C3 linearisation algorithm:
https://en.wikipedia.org/wiki/C3_linearization

**Comment by mrocklin** [comment]
Sorry, I'm unlikely to read that article in depth.  Can I ask you to summarize how it breaks ties when there isn't a clear precedence?  I don't think we care about the entire algorithm here (multipledispatch is a different situation).  I think that we care about the specific question of tiebreaking when neither class is more specific than the other.

**Comment by oscarbenjamin** [comment]
Essentially it all comes down to the order of superclasses in the class statement:
```python
In [7]: class A: pass

In [8]: class B: pass

In [9]: class C(A, B): pass

In [10]: class D(B, A): pass

In [11]: C.mro()
Out[11]: [__main__.C, __main__.A, __main__.B, object]

In [12]: D.mro()
Out[12]: [__main__.D, __main__.B, __main__.A, object]
```

**Comment by mrocklin** [comment]
Ah, I see, I was missing that context.  Sure, if you're able to make a change that respects the ordering in the provided class I'm cool with that.  We'll want to make sure that it doesn't negatively impact any of the other consdierations or affect performance.   Thank you for taking the time to explain things.  I appreciate it.

**Comment by oscarbenjamin** [comment]
I haven't looked much at the internals of multipledispatch but possibly a difficulty in implementing this is that it could only be resolved at call time rather than at decorator time e.g. it is not until `f(C())` is called that we would ever know that a class that subclasses both `A` and `B` might exist so a decision about how to handle that case could not be made until the function call itself (and therefore could not be precomputed although it could be cached).

**Comment by mrocklin** [comment]
I agree that that sounds like a reasonable concern.  My understanding of this library is out-of-date, but I suspect that this is a common concern, and that a lot of our performance depends on caching.
